### PR TITLE
[JENKINS-54346] && [JENKINS-55505] Fix stream handling in tee (followup)

### PR DESCRIPTION
**Pull request by PhilippHomann**
This is just a followup of #62 
Most credits of changes going out to @tolnaisz

Just fixed the test error mentioned in #62

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/fs/TeeStep.java [hunk 1]
```diff
@@ -97,26 +119,57 @@ public boolean start() throws Exception {
...
+            /*
+              This gets called either to serialize to allow resume or to transfer to a remote.
+              When resuming, the stream gets recreated as and when needed.
+              When transferring the stream needs to transfer as well.
+
+              The transferredToRemote is used to detect whether the stream got serialized as well
+              which is also guaranteeing backward compatibility when resuming after plug-in upgrade.
+             */
+            transferredToRemote = Channel.current() != null;
+            oos.defaultWriteObject();
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/fs/TeeStep.java:144]: In the original PR I suggested that this be rewritten to use `writeReplace` for legibility and maintainability, but I do not consider it a blocker.

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/fs/TeeStep.java [hunk 2]
```diff
@@ -97,26 +119,57 @@ public boolean start() throws Exception {
...
+                // Transferred, so the stream got transferred as well.
+                stream = (OutputStream) ois.readObject();
             }
-        } else {
+        }
+
+        private Object readResolve() {
+            return this;
+        }
+
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/fs/TeeStep.java:167]:
```suggestion
```

I guess?

File: src/test/java/org/jenkinsci/plugins/pipeline/utility/steps/fs/TeeStepTest.java [hunk 3]
```diff
@@ -74,12 +76,55 @@ public void smokes() throws Exception {
...
+                r.assertBuildStatus(Result.SUCCESS, r.waitForCompletion(b));
                 assertThat(r.getLog(b), stringContainsInOrder("got: first message second message", Functions.isWindows() ? "WS>rem" : "+ true"));
 
         });
     }
 
+    @Test
+    @Issue({"JENKINS-54346", "JENKINS-55505"})
+    public void closed() throws Exception {
+        rr.then(r -> {
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/pipeline/utility/steps/fs/TeeStepTest.java:88]: Still no test coverage of behavior across restarts that I can tell.

Comment (PhilippHomann) [reply src/test/java/org/jenkinsci/plugins/pipeline/utility/steps/fs/TeeStepTest.java:88]: See my last commit.
Is that what you expect?

Comment (jglick) [reply src/test/java/org/jenkinsci/plugins/pipeline/utility/steps/fs/TeeStepTest.java:88]: Something like that, yes. I would have to immerse myself in this patch for a while to really remember what the tricky cases were.

**Pull request merged**